core/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java

```diff
@@ -207,6 +211,24 @@ void execute(Terminal terminal, String pluginId, boolean isBatch, Map<String, St
             Files.createDirectory(env.pluginsFile());
         }
 
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            for (final Path pathToDeleteOnShutdown : pathsToDeleteOnShutdown) {
+                try {
+                    IOUtils.rm(pathToDeleteOnShutdown);
+                } catch (final IOException e) {
+                    try (
+                        final StringWriter sw = new StringWriter();
+                        final PrintWriter pw = new PrintWriter(sw)) {
+                        e.printStackTrace(pw);
+                        terminal.println(sw.toString());
+                    } catch (final IOException impossible) {
+                        // StringWriter#close declared a checked IOException from the Closeable interface but the Javadocs for StringWriter
```

Comment:
That's fair. Since it's a violation of our assumptions no matter how we get there, I opted to throw an `AssertionError`. I pushed a1a7e7e5d7960f107ce6cdaaa645b1ba1da9b9a6.